core/store/src/trie/trie_storage.rs

```diff
@@ -128,35 +131,61 @@ impl TrieCachingStorage {
     /// StorageError::StorageInternalError if the storage fails internally.
     pub fn retrieve_rc(&self, hash: &CryptoHash) -> Result<u32, StorageError> {
         let mut guard = self.cache.lock().expect(POISONED_LOCK_ERR);
-        if let Some(val) = (*guard).cache_get(hash) {
+        if let Some(val) = guard.cache_get(hash) {
             Self::vec_to_rc(val)
         } else {
             let key = Self::get_key_from_shard_id_and_hash(self.shard_id, hash);
             let val = self
                 .store
                 .get(ColState, key.as_ref())
                 .map_err(|_| StorageError::StorageInternalError)?;
-            let rc = Self::vec_to_rc(&val);
-            (*guard).cache_set(*hash, val);
-            rc
+            if let Some(val) = val {
+                let rc = Self::vec_to_rc(&val);
+                if val.len() < TRIE_MAX_VALUE_CACHED {
+                    guard.cache_set(*hash, val);
+                }
+                rc
+            } else {
+                Ok(0)
+            }
+        }
+    }
+
+    pub fn update_cache(&self, ops: Vec<(CryptoHash, Option<Vec<u8>>)>) {
+        let mut guard = self.cache.lock().expect(POISONED_LOCK_ERR);
+        for (hash, value) in ops {
+            if let Some(value) = value {
+                if value.len() < TRIE_MAX_VALUE_CACHED && guard.cache_get(&hash).is_some() {
```

Comment:
why do we require that the hash should already exist in the cache?

Reply:
The idea is that updating the cache on writes is not very useful, because in case of GC these values can be old. Ideally this cache_set should also not bump the item in priority but that's not possible with the API.

Reply:
I see. Makes sense.